### PR TITLE
feat(anoncreds): add anoncreds registry service

### PR DESCRIPTION
**Pull request by TimoGlastra**
Fixes https://github.com/hyperledger/aries-framework-javascript/issues/1123, and adds the initial structure for the `AnonCredsModule`

### CODE REVIEW

**Comment by TimoGlastra** [comment]
Thinking whether we should rename `supportedMethods` to `supportedIdentifiers` in the AnonCredsRegistry interface, as we're not really filtering based on method, but rather on the structure of the identifier

File: packages/anoncreds/src/AnonCredsModuleConfig.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @public
+ * AnonCredsModuleConfigOptions defines the interface for the options of the AnonCredsModuleConfig class.
+ */
+export interface AnonCredsModuleConfigOptions {
+  /**
+   * A list of AnonCreds registries to make available to the AnonCreds module.
+   */
+  registries: [AnonCredsRegistry, ...AnonCredsRegistry[]]
```
Comment (blu3beri) [inline packages/anoncreds/src/AnonCredsModuleConfig.ts:11]:
```suggestion
  registries: Array<AnonCredsRegistry>
```

Or am I missing a sick hidden TypeScript feature?

File: packages/anoncreds/src/services/registry/AnonCredsRegistry.ts [hunk 2]
```diff
@@ -8,8 +8,14 @@ import type { GetRevocationRegistryDefinitionReturn } from './RevocationRegistry
...
 import type { AgentContext } from '@aries-framework/core'
 
-// This service can be registered multiple times in a single AFJ instance.
+/**
+ * @public
+ */
 export interface AnonCredsRegistry {
+  // TODO: should this be a single regex? So just supportedMethodRegex. If you support multiple / has complex queries you can just combine them.
+  // This would prevent the need for a double loop (loop through all registries, then loop through all supportedMethods)
+  supportedMethods: RegExp[]
```
Comment (blu3beri) [inline packages/anoncreds/src/services/registry/AnonCredsRegistry.ts:17]: I think here we should assume the most common use case, which I think is one identifier. Which regex it is rather easy to match on `A` or `B` (`(A|B)`).

Comment (blu3beri) [reply packages/anoncreds/src/services/registry/AnonCredsRegistry.ts:17]: I think I'd prefer the `supportedIdentifier`, but I might prefer it if it indicates a regex. so `identifierPattern` or something? And we can provide a util function with would be something like this:

```typescript
const getSimpleMethodPattern(method:string): RegExp {
  return new RegExp(`^did:${method}:.*$`) // not a perfect pattern but you get the gist :).
}

File: packages/anoncreds/src/services/registry/AnonCredsRegistryService.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * @internal
+ * The AnonCreds registry service manages multiple {@link AnonCredsRegistry} instances
+ * and returns the correct registry based on a given identifier
+ */
+@injectable()
+export class AnonCredsRegistryService {
+  public async getRegistryForIdentifier(agentContext: AgentContext, identifier: string): Promise<AnonCredsRegistry> {
+    const registries = agentContext.dependencyManager.resolve(AnonCredsModuleConfig).registries
+
+    // TODO: should we check if multiple are registered?
```
Comment (blu3beri) [inline packages/anoncreds/src/services/registry/AnonCredsRegistryService.ts:19]: Ah yeah that's a good question. I am okay with picking the first one, but we have to mark it explicitly in the code documentation for the public function which uses this `getRegistryForIdentifier` under-the-hood.

File: packages/indy-sdk/src/anoncreds/services/IndySdkAnonCredsRegistry.ts [hunk 4]
```diff
@@ -33,32 +36,39 @@ import {
...
 export class IndySdkAnonCredsRegistry implements AnonCredsRegistry {
-  private indySdk: IndySdk
-  private indySdkPoolService: IndySdkPoolService
-
-  public constructor(@inject(IndySdkSymbol) indySdk: IndySdk, indySdkPoolService: IndySdkPoolService) {
-    this.indySdk = indySdk
-    this.indySdkPoolService = indySdkPoolService
-  }
+  /**
+   * This class only supports resoling and registering objects with legacy indy identifiers.
```
Comment (blu3beri) [inline packages/indy-sdk/src/anoncreds/services/IndySdkAnonCredsRegistry.ts:40]:
```suggestion
   * This class only supports resolving and registering objects with legacy indy identifiers.
```

**Review by blu3beri — CHANGES_REQUESTED** [review]
A couple of notes, nothing major.

File: packages/anoncreds/src/AnonCredsModuleConfig.ts (hunk 1 above)
Comment (TimoGlastra) [reply packages/anoncreds/src/AnonCredsModuleConfig.ts:11]: It means the array must have at least one entry in it. As using the module without registries isn't really useful

File: packages/anoncreds/src/services/registry/AnonCredsRegistry.ts (hunk 2 above)
Comment (TimoGlastra) [reply packages/anoncreds/src/services/registry/AnonCredsRegistry.ts:17]: I like it!

File: packages/anoncreds/src/AnonCredsModuleConfig.ts (hunk 1 above)
Comment (blu3beri) [reply packages/anoncreds/src/AnonCredsModuleConfig.ts:11]: Ah, cool!

**Review by blu3beri — APPROVED** [review]

**Pull request merged**